### PR TITLE
*: implement read index service

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

## What have you changed? (mandatory)

Please **NOTE** that:
- Do not assume reviewers understand the original issue

## What are the type of the changes? (mandatory)
Details in: https://github.com/tikv/rfcs/pull/14
Actually, there is an inconsistent issue about apply status between leader and follower, but we now only reply after successfully apply and write in the leader.

## The currently defined type
- New feature

## How has this PR been tested? (mandatory)
-Unit Test

### CODE REVIEW

**Comment by siddontang** [comment]
> Actually, there is an inconsistent issue about apply status between leader and follower, but we now only reply after successfully apply and write in the leader.

where is the origin issue?

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1203,24 +1225,69 @@ impl Peer {
...
+                    cb.invoke_read(self.handle_read(ctx, req, true, read_index));
+                }
+                self.pending_reads.ready_cnt -= 1;
+            }
+        }
+    }
+
     fn apply_reads<T, C>(&mut self, ctx: &mut PollContext<T, C>, ready: &Ready) {
         let mut propose_time = None;
+        // The follower may lost `ReadIndexResp`, so the pending_reads is not
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: A network may also drop `ReadIndexRequest` silently, should we retry read index after few heartbeats timeouts?

Comment (nolouch) [reply src/raftstore/store/peer.rs]: If the next request succeeds，the before request also can be processed even which may be lost in `Ready`. and the client better set the timeout for the request.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -1190,22 +1212,63 @@ impl Peer {
...
```
Comment (Connor1996) [inline src/raftstore/store/peer.rs]:
```suggestion
        // The follower may lost `ReadIndexResp`, so the pending_reads doesn't
```

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1190,22 +1212,63 @@ impl Peer {
...
+                        && read.cmds[0].0.get_requests().len() == 1
+                        && read.cmds[0].0.get_requests()[0].get_cmd_type() == CmdType::ReadIndex
+                    {
+                        (read.read_index, true)
+                    } else {
+                        (read.read_index, false)
+                    }
+                };
+
+                if !self.ready_to_handle_read() && !is_read_index_request {
```
Comment (Connor1996) [inline src/raftstore/store/peer.rs]: seems it should be or, not and?

**Pull request merged**